### PR TITLE
SWATCH-1571: Skip syncing SaaS-contract subscriptions from IT Subscription Service

### PR DESCRIPTION
**Pull request by kahowell**
Jira issue: [SWATCH-1571](https://issues.redhat.com/browse/SWATCH-1571)

Description
===========

Subscription syncing currently fetches information only from IT Subscription Service and IT Product Service. (It does not consult the partner subscription API). Contract capacities are recorded only the partner subscription API, so the view from IT Subscription Service is incomplete. This can cause a SaaS-contract subscription to be incorrectly modified during a subscription sync.

Note that the `Subscription.equals` method prevents update of a `Subscription` entity unless one of the following properties changes:

* `subscriptionNumber`
* `quantity`
* `startDate`
* `endDate`
* `billingProviderId`
* `billingAccountId`
* `accountNumber`
* `billingProvider`

Contract service does not record accountNumber, so when the subscription is synced, it'll be considered unequal (null != non-null), and then it'll remove the subscription measurements.

Testing
=======

1. Deploy the service:

```shell
DEV_MODE=true \
  DEVTEST_EVENT_EDITING_ENABLED=true \
  SUBSCRIPTION_USE_STUB=true \
  PRODUCT_USE_STUB=true \
  ./gradlew :bootRun
```

2. Sync subscriptions:

```shell
http PUT :8000/api/rhsm-subscriptions/v1/internal/subscriptions/sync/org/123 \
  x-rh-swatch-psk:placeholder \
  Origin:console.redhat.com
```

3. Update the subscription to look like a SaaS-contract subscription, update capacity too:

```shell
psql -h localhost -U rhsm-subscriptions rhsm-subscriptions <<EOF
update subscription set billing_provider='aws' where id='235255';
update subscription_measurements set value=100 where id='235255';
EOF
```

4. Sync subscriptions:

```shell
http PUT :8000/api/rhsm-subscriptions/v1/internal/subscriptions/sync/org/123 \
  x-rh-swatch-psk:placeholder \
  Origin:console.redhat.com
```

Verification
------------

See in the logs a line like:
> Skipping sync of subscriptionId=235255 because it has contract-provided capacity

Verify the amounts are unchanged:

```shell
psql -h localhost -U rhsm-subscriptions rhsm-subscriptions <<EOF
select * from subscription_measurements where subscription_id='235255';
EOF
```

### CODE REVIEW

File: src/main/java/org/candlepin/subscriptions/subscription/SubscriptionSyncController.java [hunk 1]
```diff
@@ -219,6 +214,16 @@ public void syncSubscription(
       final org.candlepin.subscriptions.db.model.Subscription existingSubscription =
           subscriptionOptional.get();
       log.debug("Existing subscription in DB={}", existingSubscription);
+      if (Objects.nonNull(existingSubscription.getBillingProvider())
+          && !existingSubscription.getSubscriptionMeasurements().isEmpty()) {
```
[the remark this answers is not in the document]
Comment (kahowell) [reply src/main/java/org/candlepin/subscriptions/subscription/SubscriptionSyncController.java:218]: As discussed in person, I prefer a more comprehensive fork here for now; that will reduce the risk of regressions here until we have a chance to reconcile the subscription syncing with contract syncing more comprehensively.

**Review by kartikshahc — APPROVED** [review]

**Comment by himmAllRight** [comment]
Spun up an EE against the latest build of this branch, and from what I can tell, the capacity report being set from the contract looks good:

```python
In [42]: start, end=get_range(days=5)

In [43]: (start,end)
Out[43]: ('2023-08-23T00:00:00Z', '2023-08-28T18:57:13.620979Z')

In [44]: m = [{'metric_id': 'four_vcpu_0', 'value': 250},
    ...:  {'metric_id': 'control_plane_0', 'value': 250}]

In [45]: app.rhsm_subscriptions.create_contract('rosa', contract_sku="MW01484", start_date="2023-08-25T16:37:54Z", contract_info={'metrics': m})
2023-08-28 18:57:49.357 [    INFO] [root] Starting POST /internal/contracts in clowder_smoke env ...
2023-08-28 18:57:49.395 [    INFO] [root]

request response url: http://swatch-contracts-service:8000/api/swatch-contracts/internal/contracts, status_code: 200
2023-08-28 18:57:49.396 [    INFO] [root] Respone return: {"billing_account_id":"f7543e2b-ec9f-472c-9ff6-d84d26554b11","billing_provider":"aws","metrics":[{"metric_id":"four_vcpu_0","value":250},{"metric_id":"control_plane_0","value":250}],"org_id":"3340851","product_id":"rosa","sku":"MW01484","start_date":"2023-08-2
5T16:37:54Z","subscription_number":"8524db61-90cb-4c62-a1bc-a489ae7e5f94","uuid":"784015f6-7961-4bb0-b08a-a9f6b4af3b6f","vendor_product_code":"8f7174af-186b-4c1e-b90e-ebd484fbd365"}
2023-08-28 18:57:49.396 [    INFO] [root] {'billing_account_id': 'f7543e2b-ec9f-472c-9ff6-d84d26554b11', 'billing_provider': 'aws', 'metrics': [{'metric_id': 'four_vcpu_0', 'value': 250}, {'metric_id': 'control_plane_0', 'value': 250}], 'org_id': '3340851', 'product_id': 'rosa', 'sku': 'MW01484', 'start_date': '2023-
08-25T16:37:54Z', 'subscription_number': '8524db61-90cb-4c62-a1bc-a489ae7e5f94', 'uuid': '784015f6-7961-4bb0-b08a-a9f6b4af3b6f', 'vendor_product_code': '8f7174af-186b-4c1e-b90e-ebd484fbd365'}
Out[45]:
{'billing_account_id': 'f7543e2b-ec9f-472c-9ff6-d84d26554b11',
 'billing_provider': 'aws',
 'metrics': [{'metric_id': 'four_vcpu_0', 'value': 250},
  {'metric_id': 'control_plane_0', 'value': 250}],
 'org_id': '3340851',
 'product_id': 'rosa',
 'sku': 'MW01484',
 'start_date': '2023-08-25T16:37:54Z',
 'subscription_number': '8524db61-90cb-4c62-a1bc-a489ae7e5f94',
 'uuid': '784015f6-7961-4bb0-b08a-a9f6b4af3b6f',
 'vendor_product_code': '8f7174af-186b-4c1e-b90e-ebd484fbd365'}

In [46]: app.rhsm_subscriptions.get_capacity_report(product_id="rosa", metric_id="Instance-hours",beginning=start,ending=end,granularity="Daily")
2023-08-28 18:57:57.116 [    INFO] [root] REST: http://swatch-api-service.ephemeral-jfkvob.svc:8000/api/rhsm-subscriptions/v1/capacity/products/rosa/Instance-hours
2023-08-28 18:57:57.144 [    INFO] [iqe.base.rest_client] REST: METHOD=GET, request_id=None, params=[('granularity', 'Daily'), ('beginning', '2023-08-23T00:00:00+00:00'), ('ending', '2023-08-28T18:57:13.620979+00:00')]
Out[46]:
{'data': [{'date': datetime.datetime(2023, 8, 23, 0, 0, tzinfo=tzlocal()),
           'has_data': False,
           'has_infinite_quantity': False,
           'value': 0},
          {'date': datetime.datetime(2023, 8, 24, 0, 0, tzinfo=tzlocal()),
           'has_data': False,
           'has_infinite_quantity': False,
           'value': 0},
          {'date': datetime.datetime(2023, 8, 25, 0, 0, tzinfo=tzlocal()),
           'has_data': False,
           'has_infinite_quantity': False,
           'value': 0},
          {'date': datetime.datetime(2023, 8, 26, 0, 0, tzinfo=tzlocal()),
           'has_data': True,
           'has_infinite_quantity': False,
           'value': 250},
          {'date': datetime.datetime(2023, 8, 27, 0, 0, tzinfo=tzlocal()),
           'has_data': True,
           'has_infinite_quantity': False,
           'value': 250},
          {'date': datetime.datetime(2023, 8, 28, 0, 0, tzinfo=tzlocal()),
           'has_data': True,
           'has_infinite_quantity': False,
           'value': 250}],
 'meta': {'count': 6,
          'granularity': 'Daily',
          'metric_id': 'Instance-hours',
          'product': 'rosa'}}

In [47]: app.rhsm_subscriptions.get_capacity_report(product_id="rosa", metric_id="Cores",beginning=start,ending=end,granularity="Daily")
2023-08-28 18:58:18.714 [    INFO] [root] REST: http://swatch-api-service.ephemeral-jfkvob.svc:8000/api/rhsm-subscriptions/v1/capacity/products/rosa/Cores
2023-08-28 18:58:18.745 [    INFO] [iqe.base.rest_client] REST: METHOD=GET, request_id=None, params=[('granularity', 'Daily'), ('beginning', '2023-08-23T00:00:00+00:00'), ('ending', '2023-08-28T18:57:13.620979+00:00')]
Out[47]:
{'data': [{'date': datetime.datetime(2023, 8, 23, 0, 0, tzinfo=tzlocal()),
           'has_data': False,
           'has_infinite_quantity': False,
           'value': 0},
          {'date': datetime.datetime(2023, 8, 24, 0, 0, tzinfo=tzlocal()),
           'has_data': False,
           'has_infinite_quantity': False,
           'value': 0},
          {'date': datetime.datetime(2023, 8, 25, 0, 0, tzinfo=tzlocal()),
           'has_data': False,
           'has_infinite_quantity': False,
           'value': 0},
          {'date': datetime.datetime(2023, 8, 26, 0, 0, tzinfo=tzlocal()),
           'has_data': True,
           'has_infinite_quantity': False,
           'value': 1000},
          {'date': datetime.datetime(2023, 8, 27, 0, 0, tzinfo=tzlocal()),
           'has_data': True,
           'has_infinite_quantity': False,
           'value': 1000},
          {'date': datetime.datetime(2023, 8, 28, 0, 0, tzinfo=tzlocal()),
           'has_data': True,
           'has_infinite_quantity': False,
           'value': 1000}],
 'meta': {'count': 6,
          'granularity': 'Daily',
          'metric_id': 'Cores',
          'product': 'rosa'}}
```

However, I'm having some issues verifying the sync skip and message in EE, mostly because I'm struggling to get the account to be setup correctly in EE to verify that case..

**Comment by himmAllRight** [comment]
@kahowell I'm good to have this merge and finish the remaining verification in stage. This issue should also be covered by the SWatch-1526 automation I'm working on, which is blocked by this issue (so it should have automated coverage coming shortly)

**Review by himmAllRight — APPROVED** [review]

**Pull request merged**